plugins/flatpak/gs-flatpak.c

```diff
@@ -2905,6 +2905,78 @@ gs_flatpak_add_app_to_list_if_not_installed (GsApp *app, GsAppList *list,
 	gs_app_list_add (list, app);
 }
 
+static GPtrArray *
+gs_flatpak_get_app_remote_related_refs (GsFlatpak *self,
+					GsApp *app,
+					GCancellable *cancellable,
+					GError **error)
+{
+	g_autofree gchar *ref = NULL;
+
+	/* lookup any related refs for this ref */
+	ref = g_strdup_printf ("%s/%s/%s/%s",
+			       gs_flatpak_app_get_ref_kind_as_str (app),
+			       gs_flatpak_app_get_ref_name (app),
+			       gs_flatpak_app_get_ref_arch (app),
+			       gs_flatpak_app_get_ref_branch (app));
+	return flatpak_installation_list_remote_related_refs_sync (self->installation,
+								   gs_app_get_origin (app),
+								   ref, cancellable, error);
+}
+
+static gboolean
+gs_flatpak_add_related_refs_to_list (GsFlatpak *self,
+				     GsApp *app,
+				     GPtrArray *related_refs,
+				     GHashTable *hash_installed,
+				     gboolean is_update,
+				     GsAppList *list,
+				     GCancellable *cancellable,
+				     GError **error)
+{
+	for (guint i = 0; i < related_refs->len; i++) {
+		FlatpakRelatedRef *xref_related = g_ptr_array_index (related_refs, i);
+		g_autofree gchar *ref_display = NULL;
+		g_autoptr(GsApp) app_tmp = NULL;
+		g_autoptr(GError) error_local = NULL;
+
+		/* not included */
+		if (!gs_flatpak_related_should_download (self, app, xref_related))
+			continue;
+
+		/* already installed? */
+		app_tmp = gs_flatpak_create_app (self, FLATPAK_REF (xref_related));
+		ref_display = gs_flatpak_app_get_ref_display (app_tmp);
+		if (!is_update && g_hash_table_contains (hash_installed, ref_display)) {
+			g_debug ("not adding related %s as already installed", ref_display);
+			continue;
+		}
+
+		if (gs_app_get_origin (app_tmp) == NULL)
+			gs_app_set_origin (app_tmp, gs_app_get_origin (app));
```

Comment:
The old code was doing this unconditionally; any reason you changed it?

Reply:
Yes. I have noticed that it gave a warning when the `app_tmp` already has an origin set. So I sneaked it in since the whole block was changed.